sklearn/linear_model/coordinate_descent.py

```diff
@@ -696,28 +732,70 @@ def fit(self, X, y, check_input=True):
+
         if isinstance(self.precompute, str):
             raise ValueError('precompute should be one of True, False or'
                              ' array-like. Got %r' % self.precompute)
 
+        all_solvers = ['auto', 'cd', 'saga']
+        if self.solver not in all_solvers:
+            raise ValueError("ElasticNet Regression supports only solvers in"
+                             " {}, got {}.".format(all_solvers, self.solver))
+
+        if self.solver == 'auto':
+            if (isinstance(X, np.ndarray) and X.flags['F_CONTIGUOUS']) or \
+             (sparse.issparse(X) and sparse.isspmatrix_csc(X)):
+                solver = 'cd'
+            else:
+                solver = 'saga'
+        else:
+            solver = self.solver
+
+        if self.positive and solver == 'saga':
+            raise ValueError("ElasticNet supports positive contraints for "
+                             "coefficients only for solver 'cd', "
+                             "got solver={}, positive={}"
+                             .format(solver, self.positive))
+
         # Remember if X is copied
         X_copied = False
-        # We expect X and y to be float64 or float32 Fortran ordered arrays
-        # when bypassing checks
+        # For 'cd', we expect X and y to be float64 or float32 Fortran ordered
+        # arrays when bypassing checks
+        # For 'saga' we expect float64 C ordered arrays
         if check_input:
             X_copied = self.copy_X and self.fit_intercept
-            X, y = check_X_y(X, y, accept_sparse='csc',
-                             order='F', dtype=[np.float64, np.float32],
-                             copy=X_copied, multi_output=True, y_numeric=True)
-            y = check_array(y, order='F', copy=False, dtype=X.dtype.type,
-                            ensure_2d=False)
+            if solver == 'cd':
+                X, y = check_X_y(X, y, accept_sparse='csc',
+                                 order='F', dtype=[np.float64, np.float32],
+                                 copy=X_copied, multi_output=True,
+                                 y_numeric=True)
+                y = check_array(y, order='F', copy=False, dtype=X.dtype.type,
+                                ensure_2d=False)
+            else:
+                X, y = check_X_y(X, y, accept_sparse='csr',
+                                 order='C', dtype=np.float64,
+                                 copy=X_copied, multi_output=True,
+                                 y_numeric=True)
+                # For 2D target, we need contiguous columns as they are
+                # to be regressed independently on X
+                y = check_array(y, order='F', copy=False, dtype=X.dtype.type,
+                                ensure_2d=False)
 
         # Ensure copying happens only once, don't do it again if done above
         should_copy = self.copy_X and not X_copied
+        if solver == 'cd':
+            precompute_ = self.precompute
```

Comment:
You raise a `ValueError` above for an incompatible `positive` argument, but here you silently fix it. Shouldn't these two follow the same logic? (either raise error or silently ignore).

Reply:
You don't get what you want or expect when you set ``positive=True`` and use saga. They are truly incompatible, therefore I raise ``ValueError``.
On the other side, if you specify ``precompute=True`` (or an array) and use saga, you get correct results. Saga does just not use the info from ``precompute``. 
Is there a way to make a distinction between arguments on which the result depends (``alpha``,  ``fit_intercept``, ``positive``), and arguments for optimization (tuning) (``solver``, ``precompute``, ``warm_start``)?

Reply:
I see. I would still not be too unhappy with a `ValueError`, but in this case a `UserWarning` could also do the job I guess.

Reply:
Good idea. Done.